inappmessaging/src/main/java/com/rakuten/tech/mobile/inappmessaging/runtime/data/repositories/CampaignRepository.kt

```diff
@@ -0,0 +1,208 @@
+package com.rakuten.tech.mobile.inappmessaging.runtime.data.repositories
+
+import androidx.annotation.VisibleForTesting
+import com.google.gson.Gson
+import com.rakuten.tech.mobile.inappmessaging.runtime.InAppMessaging
+import com.rakuten.tech.mobile.inappmessaging.runtime.data.models.messages.Message
+import com.rakuten.tech.mobile.inappmessaging.runtime.data.responses.ping.CampaignData
+import com.rakuten.tech.mobile.sdkutils.PreferencesUtil
+import com.rakuten.tech.mobile.sdkutils.logger.Logger
+import org.json.JSONObject
+import java.lang.ClassCastException
+import java.lang.Integer.max
+
+internal interface CampaignRepositoryType {
+    val messages: LinkedHashMap<String, Message>
+
+    val lastSyncMillis: Long?
+
+    /**
+     * Syncs [messages] with server.
+     */
+    fun syncWith(messageList: List<Message>, timestampMillis: Long)
+
+    /**
+     * Updates the [Message.isOptedOut] as true for the provided campaign.
+     */
+    fun optOutCampaign(campaign: Message): Message?
+
+    /**
+     * Decrements the number of [Message.impressionsLeft] for provided campaign Id in the repository.
+     */
+    fun decrementImpressions(id: String): Message?
+
+    /**
+     * Increments the number of [Message.impressionsLeft] for provided campaign Id in the repository.
+     */
+    fun incrementImpressions(id: String): Message?
+
+    /**
+     * Loads campaign data from user cache.
+     */
+    fun loadCachedData()
+
+    /**
+     * Clears [messages] for last user.
+     */
+    fun clearMessages()
+}
+
+@SuppressWarnings("UnnecessaryAbstractClass")
+internal abstract class CampaignRepository : CampaignRepositoryType {
+    companion object {
+        private var instance: CampaignRepository = CampaignRepositoryImpl()
+
+        @VisibleForTesting
+        internal const val IAM_USER_CACHE = "IAM_user_cache"
+        private const val TAG = "IAM_CampaignRepo"
+
+        fun instance(): CampaignRepository = instance
+    }
+
+    private class CampaignRepositoryImpl : CampaignRepository() {
+
+        // LinkedHashMap can preserve the message insertion order.
+        // Map - Key: Campaign ID, Value: Message object
+        private val messagesHashMap = LinkedHashMap<String, Message>()
+
+        init {
+            loadCachedData()
+        }
+
+        override val messages: LinkedHashMap<String, Message>
```

Comment:
Why is it needed to have `messages` and `messagesHashMap`?

Reply:
Initially the purpose of `messages` is to provide a copy of `messagesHashmap` so that oldList doesn't get cleared too (line 81-83)
I understood the comment now, I updated to only 1 variable pertaining to the list `messages`